File: Solutions/563_Binary_Tree_Tilt/563_Binary_Tree_Tilt.go
//Second fastest
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findTilt(root *TreeNode) int {
    _, res := tilt(root)
    return res
}

func tilt(root *TreeNode) (int,int) {
    if root == nil {
        return 0, 0
    }
    left, s1 := tilt(root.Left)
    right, s2 := tilt(root.Right)
    return (root.Val+left+right), (abs(left-right)+s1+s2)
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

//Another
import "math"
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findTilt(root *TreeNode) int {
    tilt := 0
    helperFindTilt(root,&tilt)
    return tilt
}

func helperFindTilt(root *TreeNode, tilt *int) int {
    if root == nil {
        return 0
    }
    
    if root.Left == nil && root.Right == nil {
        *tilt += 0
        return root.Val
    }
    
    left := helperFindTilt(root.Left,tilt)
    right := helperFindTilt(root.Right,tilt)
    
    *tilt += int(math.Abs(float64(left - right)))
    return left+ right + root.Val
}

//Fastest
func findTilt(root *TreeNode) int {
	result := 0
	itterate(root, &result)
	return result
}

func itterate(node *TreeNode, result *int) int {
	if node == nil {
		return 0
	}

	left, right := itterate(node.Left, result), itterate(node.Right, result)
	*result += absDiff(left, right)
	return left + right + node.Val
}

func absDiff(a, b int) int {
	diff := a - b
	if diff < 0 {
		diff = -diff
	}

	return diff
}

File: Solutions/563_Binary_Tree_Tilt/563_Binary_Tree_Tilt.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int FindTilt(TreeNode root) {
        int tilt = 0;
        Traverse(root, ref tilt);
        return tilt;
    }
    
    public int Traverse(TreeNode root, ref int tilt){
        if(root == null)
            return 0;
        
        int left = Traverse(root.left, ref tilt);
        int right = Traverse(root.right, ref tilt);
        
        tilt += Math.Abs(left - right);
        
        return left + right + root.val;
    }
}

//Second fastest
public class Solution {
    public int FindTilt(TreeNode root) {
        var t = GetTilt(root);
        return t.Item2;
    }
    
    private Tuple<int, int> GetTilt(TreeNode root) {
        if(root == null) return new Tuple<int, int>(0, 0);
        var left = GetTilt(root.left);
        var right = GetTilt(root.right);
        var tilt = Math.Abs(left.Item1 - right.Item1) + left.Item2 + right.Item2;
        var sum = left.Item1 + right.Item1 + root.val;
        return new Tuple<int, int>(sum, tilt);
    }
}

//Fastest
public class Solution {
    int res = 0; 
    public int FindTilt(TreeNode root) {
        postOrder(root);
        return res; 
            
        
    }
    
    private int postOrder(TreeNode node){
        
        if(node == null )
            return 0;
        
        int left = postOrder(node.left); 
        int right = postOrder(node.right); 
        
        res += Math.Abs(left - right) ; 
        
        return node.val + right + left; 
    }
}